### PR TITLE
Bug 916971 - Write npm-mirror utility in nodejs r=wlach

### PR DESCRIPTION
**Pull request by lambdabaa**

### CODE REVIEW

File: lib/server.js [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+  var stream = fs.createWriteStream(loc);
+  debug('GET ' + url);
+  http.get(url, function(res) {
+    if (res.statusCode !== 200) {
+      callback(new Error('Bad status ' + res.statusCode + ' for ' + url));
+    }
+
+    res.pipe(stream);
+    stream.on('finish', function() {
+      stream.close();
```
Comment (lightsofapollo) [inline lib/server.js]: why do you need to call close directly?

File: lib/server.js [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+   * Sync the named package.
+   *
+   * @param {string} packageName name of package to sync.
+   * @param {Function} callback invoke when done.
+   */
+  syncPackage: function(packageName, callback) {
+    debug('sync ' + packageName);
+    var packageUrl = url.resolve(this.registry, packageName);
+    debug('GET ' + packageUrl);
+    http.get(packageUrl, function(res) {
```
Comment (lightsofapollo) [inline lib/server.js]: this seems really expensive in terms of a network call (and its io bound on the disk too)... Does the server support etags? We can dramatically cut costs by doing a  If-None-Match request (or maybe its If-Match I get them mixed up sometimes =p)

File: lib/server.js [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+      if (res.statusCode !== 200) {
```
Comment (lightsofapollo) [inline lib/server.js]: if you take the above approach checking > 199 and < then 300 is the right thing

File: lib/server.js [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+      });
+      res.on('end', function() {
+        var packagePath = path.resolve(this.packageDir, packageName);
+        if (!fs.existsSync(packagePath)) {
+          debug('mkdir ' + packagePath);
+          fs.mkdirSync(packagePath);
+        }
+
+        var metadataFile = path.resolve(packagePath, 'index.json');
+        fs.writeFileSync(metadataFile, metadata);
```
Comment (lightsofapollo) [inline lib/server.js]: writing sync will dramatically slow down this process

File: lib/server.js [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+      res.setEncoding('utf-8');
+      var metadata = '';
+      res.on('data', function(data) {
+        metadata += data;
+      });
+      res.on('end', function() {
+        var packagePath = path.resolve(this.packageDir, packageName);
+        if (!fs.existsSync(packagePath)) {
+          debug('mkdir ' + packagePath);
+          fs.mkdirSync(packagePath);
```
Comment (lightsofapollo) [inline lib/server.js]: see below

File: lib/server.js [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+          }
+        }
+
+        for (var version in versions) {
+          this.syncPackageVersion(
+            packageName, version, metadata, onPackageSync);
+        }
+      }.bind(this));
+    }.bind(this)).on('error', function(e) {
+      throw e;
```
Comment (lightsofapollo) [inline lib/server.js]: !!! this is the default behavior of onerror anyway but you must use a callback here

File: lib/server.js [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+   * @param {string} version package version.
+   * @param {Object} metadata package metadata.
+   * @param {Function} callback invoke when done.
+   */
+  syncPackageVersion: function(packageName, version, metadata, callback) {
+    var packagePath = path.resolve(this.packageDir, packageName);
+    var tarball = metadata.versions[version].dist.tarball;
+    var tarballName = path.basename(tarball);
+    var tarballPath = path.resolve(packagePath, tarballName);
+    if (fs.exists(tarballPath)) {
```
Comment (lightsofapollo) [inline lib/server.js]: should be async if speed is at all of a concern here (and if we are writing this sync python seems like a better candidate in some ways)

File: test/unit/server_test.js [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+      sinon.assert.calledWith(syncPackage, 'a');
+      sinon.assert.calledWith(syncPackage, 'b');
+    });
+
+    test('should invoke callback once all packages are synced', function(done) {
+      syncPackage.callsArg(1);
+      subject.sync(done);
+    });
+
+    test('should throw error when received', function(done) {
```
Comment (lightsofapollo) [inline test/unit/server_test.js]: in general (its js best practice) if you have code that takes a callback and is async you must always be async and never throw (typo had to fix it!!) an exception... Always pass it in the callback.

File: lib/server.js (hunk 1 above)
Comment (lambdabaa) [inline lib/server.js]: Good point!

File: lib/server.js (hunk 4 above)
Comment (lambdabaa) [inline lib/server.js]: We can write async here I think

File: lib/server.js (hunk 6 above)
Comment (lambdabaa) [inline lib/server.js]: Done and thanks for the tip

File: test/unit/server_test.js (hunk 8 above)
Comment (lambdabaa) [inline test/unit/server_test.js]: I somehow always forget... but yes this makes sense!

File: test/unit/server_test.js [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+    test('should invoke callback once all packages are synced', function(done) {
+      syncPackage.callsArg(1);
+      subject.sync(done);
+    });
+
+    test('should pass along error', function(done) {
+      subject.packages = ['a'];
+      var err = new Error('ruhroh');
+      syncPackage.callsArgWith(1, err);
```
Comment (lightsofapollo) [inline test/unit/server_test.js]: you should probably use: `callsArgWithAsync` here

File: test/unit/server_test.js [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+    });
+
+    test('should issue a sync for every package', function() {
+      subject.sync();
+      sinon.assert.calledWith(syncPackage, 'a');
+      sinon.assert.calledWith(syncPackage, 'b');
+    });
+
+    test('should invoke callback once all packages are synced', function(done) {
+      syncPackage.callsArg(1);
```
Comment (lightsofapollo) [inline test/unit/server_test.js]: see below

File: lib/server.js (hunk 7 above)
Comment (lambdabaa) [inline lib/server.js]: Done!

File: test/unit/server_test.js (hunk 9 above)
Comment (lambdabaa) [inline test/unit/server_test.js]: Ah that's so cool. I <3 sinon

File: lib/server.js [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+      callback(new Error('Bad status ' + res.statusCode + ' for ' + url));
+    }
+
+    res.pipe(stream);
+    stream.on('finish', callback);
+  });
+};
+
+/**
+ * Make the directory iff doesn't exist.
```
Comment (lightsofapollo) [inline lib/server.js]: iff -> if it

File: lib/server.js [hunk 12]
```diff
@@ -0,0 +1,251 @@
...
+
+      var packageUrl = url.resolve(this.registry, packageName);
+      Server.download(packageUrl, function(e, data) {
+        if (e) {
+          return callback(e);
+        }
+
+        // Write metadata to fs.
+        var metadataFile = path.resolve(packagePath, 'index.json');
+        fs.writeFile(metadataFile, data);
```
Comment (lightsofapollo) [inline lib/server.js]: no callback?

File: test/unit/server_test.js [hunk 13]
```diff
@@ -0,0 +1,148 @@
+/*global __dirname, JSON, assert, path, require, setup, sinon, suite, teardown,
```
Comment (lightsofapollo) [inline test/unit/server_test.js]: __dirname and JSON? there must be some node option for jshint so you don't need to define stuff that is available in every single file.

File: lib/server.js (hunk 11 above)
Comment (lambdabaa) [inline lib/server.js]: #math

File: lib/server.js (hunk 12 above)
Comment (lambdabaa) [inline lib/server.js]: So we'll read this _eventually_ when someone tries to install the package... I guess this is a bit sketchy

File: test/unit/server_test.js (hunk 13 above)
Comment (lambdabaa) [inline test/unit/server_test.js]: Yeah I'll dig around for something like that later

File: test/helper.js [hunk 14]
```diff
@@ -0,0 +1,23 @@
+/*global global, require*/
+global.assert = require('assert');
+global.fs = require('fs');
+global.path = require('path');
+global.sinon = require('sinon');
```
Comment (jugglinmike) [inline test/helper.js:5]: We can shorten some of the tests by mixing Sinon's assertions into Node's assertion module:

``` javascript
global.sinon.assert.expose(global.assert);
```

**Comment by wlach** [comment]
So I'm not sure if I'm qualified to give coding style advice here, as I'm still relatively new to node. 

However, as far as I can tell your version of this tool doesn't seem to resolve dependencies of packages automatically, which the original python npm-mirror did (as did my prototype nodejs script at http://people.mozilla.org/~wlachance/npm-mirror.js). Without this feature, we'll need to specify all deps manually as packages to mirror, which I think is liable to break.

It's also not immediately clear to me what a "manifest" file is that you want to pass into the utility. What's wrong with just specifying the top-level packages you want to mirror on the command line? If we must specify this information in a file, I would like to see the format of it documented in the README.

Was there a reason you opted to not use my script as a starting point for your work? It may well have had a few bugs, but it seemed to be working for me when I left off.

**Comment by wlach** [comment]
Eugh, didn't mean to close the pull request, sorry.

**Comment by lambdabaa** [comment]
Chatted on irc but mirroring here for the benefit of others...

Recursive dependency resolution is a WIP but definitely needs to be here.

The manifest file is a package.json file that declares the dependencies we want.

Also, I did use your script! Just more in thought/direction than in code.

File: test/helper.js (hunk 14 above)
Comment (lambdabaa) [inline test/helper.js:5]: Cool -- will do :+1:

**Pull request merged**